feat(call-pill): add default toggled state

fabiangarga opened the pull request:
# PR Title

<!--- Feel free to remove any unused sections -->

## :hammer_and_wrench: Feat

<!--- Tick or place an `x` in all of the checkboxes that apply -->

- [ ] Fix
- [x] Feature
- [ ] Refactoring
- [ ] Documentation

## :book: Description

Add a default behavior to control the default state of the collapsible component

## :bulb: Context

For some of our implementation, there is a need to render the call pill with the default state of the collapsible component being open

## :pencil: Checklist

<!--- Tick or place an `x` in all of the checkboxes that apply -->
<!--- Remove checkboxes that do not apply -->

- [x] I have reviewed my changes
- [x] I have added tests
- [x] I have added all relevant documentation
- [ ] I have validated components with a screen reader
- [ ] I have validated components keyboard navigation
- [x] I have considered the performance impact of my change
- [x] I have checked that my change did not significantly increase bundle size
- [ ] I am exporting any new components or constants in the index.js in the component directory
- [ ] I am exporting any new components or constants in the index.js in the root

## :crystal_ball: Next Steps

<!--- Describe any future changes that need to be made after merging the PR -->

## :camera: Screenshots / GIFs

https://github.com/dialpad/dialtone-vue/assets/5950950/93a0298c-faaa-4205-9b2c-38220467d77c

fabiangarga committed:
feat(call-pill): add default toggled state

fabiangarga commented on recipes/conversation_view/feed_pill/feed_item_pill.test.js:
```diff
@@ -1,6 +1,6 @@
 import { mount } from '@vue/test-utils';
 import DtRecipeFeedItemPill from './feed_item_pill.vue';
-import { beforeEach } from 'vitest';
+import { beforeEach, describe } from 'vitest';
```
Not sure if `describe` or `beforeEach` is actually needed, noticed that runs ok without it.

juliodialpad replied:
They're not needed, please remove them.

fabiangarga commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -18,7 +18,7 @@
                 <dt-icon
                   size="300"
-                  class="dt-feed-item-pill--icon d-pr8"
+                  class="dt-feed-item-pill--icon d-pr8 d-box-content"
```
Might need to handle this in a different way, but `box-content` is required to avoid the icon shrinking when external usage has a `normalize` library that's set's pretty much everything to `border-box` that ends up in:

![image](https://github.com/dialpad/dialtone-vue/assets/5950950/233e136c-7ba6-4c47-8cf2-acb10610b826)

juliodialpad replied:
We're no longer using utility classes at component level, just for consumer side.

You might want to add that to `.dt-feed-item-pill--icon` defined below.

That image is after or before applying `box-content`?

fabiangarga replied:
Without applying the `box-content` in some apps we normalize with `* { box-sizing: border-box }` so the `svg` will have that as a value, and causes the icon to shrink.

Didn't know that we are not using utlity classes at the component level. Does this mean that `d-pr8` which is a util class should also be removed?

juliodialpad replied:
Yes, please remove `d-pr8`, we're trying to not introduce any new utility class and we're planning to make a cleanup of every component soon.

fabiangarga commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -103,6 +103,11 @@ export default {
       default: () => true,
     },
 
+    defaultToggled: {
```
Open for feedback for a better naming, `initialyToggled`?

juliodialpad replied:
You can use `isToggled` use the prefix convention `is`, `has`, `can` when you're defining boolean values.

fabiangarga replied:
Yes, my doubt with `isToggled` is that it kind of also infers the meaning that can be toggled through that, like we do with the `open` property from the `collapsible` component

juliodialpad replied:
Hmm I understand, so `defaultExpanded` might work I think

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.mdx:
```diff
@@ -18,6 +18,16 @@ import * as FeedItemPillStories from './feed_item_pill.stories';
 
 ## Variants
 
+Can enable the initial toggled state
```
```suggestion
Set the initial toggled state
```

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.mdx:
```diff
@@ -18,6 +18,16 @@ import * as FeedItemPillStories from './feed_item_pill.stories';
+  <Story of={FeedItemPillStories.DefaultToggledVariant} />
+</Canvas>
+
+Can enable the initial toggled state, and prevent to be toggled
```
```suggestion
Set the initial toggled state, and prevents toggle.
```

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.stories.js:
```diff
@@ -73,6 +73,31 @@ export const Default = {
   },
 };
 
+export const DefaultToggledVariant = {
+  render: DefaultTemplate,
+  parameters: { a11y: { disable: true } },
+  args: {
+    iconName: 'video',
+    title: 'This meeting has ended',
+    wrapperClass: 'd-w628',
+    ariaLabel: 'Click to expand',
+    toggleable: true,
+    defaultToggled: true,
+  },
+};
+
+export const DefaultToggledVariantNonToggleable = {
+  render: DefaultTemplate,
+  parameters: { a11y: { disable: true } },
+  args: {
+    iconName: 'video',
+    title: 'This meeting has ended',
+    wrapperClass: 'd-w628',
+    toggleable: false,
+    defaultToggled: true,
+  },
+};
```
Please avoid duplicating code, if you're using the same args for multiple histories, extract them into a constant and just override the needed properties.

Cleanup the objects below if you can please, they're all using the same args. check other `.stories.js` files for guidance on the best way to do that.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.test.js:
```diff
@@ -14,6 +14,7 @@ describe('DtRecipeFeedItemPill Tests', function () {
     ariaLabel: MOCK_ARIA_LABEL,
     buttonClass: '',
     toggleable: true,
+    defaultToggled: false,
```
As this is the default value, you don't need to set it here.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.test.js:
```diff
@@ -113,5 +114,31 @@ describe('DtRecipeFeedItemPill Tests', function () {
+    describe('Default toggled state close', function () {
+      beforeAll(() => {
+        mockProps = {
+          defaultToggled: false,
```
No need to mock a default value

juliodialpad reviewed (changes requested):
Please attend comments and upload the media on PR description again.

<img width="516" alt="Screenshot 2023-10-18 at 1 25 02 p m" src="https://github.com/dialpad/dialtone-vue/assets/87546543/2765cb42-c314-49f1-90f5-0be3d4cbd662">

fabiangarga committed:
fix(feed_item_pill): remove DT utils & cleanup

fabiangarga committed:
doc(feed_item_pill): update toggled wording

fabiangarga commented:
@juliodialpad thanks for the feedback, changes are good to review.
@alexdialpad are the changes ok from your side 🙏?

fabiangarga committed:
docs(feed_item_pill): move base style order

Why:
- Closer to the control
- Reactive with the controls

fabiangarga commented:
Updated the code once again, the last commit https://github.com/dialpad/dialtone-vue/pull/1261/commits/ba8738c93a99e21a265ff95373892adc590b76fa added a super small change but is disruptive with the current convention of rendering the default component at the beginning of the doc. Since it is closer to the default values that we use in the template which are the ones reactive with the control.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.mdx:
```diff
@@ -38,6 +32,12 @@ Can disable the toggleable behaviour (Cursor is reset, no changing of icon on ho
   <Story of={FeedItemPillStories.NoToggleVariant} />
 </Canvas>
 
+## Base Style
+
+<Canvas>
+  <Story of={FeedItemPillStories.Default} />
+</Canvas>
+
```
Base style should always be the first example

fabiangarga replied:
I feel like the control should be next before the variants? otherwise it feels odd to be "controlling" the props and having to scroll to see the changes in the base style.

juliodialpad replied:
You're not supposed to use controls to see changes on documentation, that's what the playground is for. Having the props there is so you can read abouth props and what they're meant for.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.stories.js:
```diff
@@ -61,66 +65,38 @@ const DefaultTemplate = (args, { argTypes }) => createTemplateFromVueFile(
-    wrapperClass: 'd-w628',
-    borderColor: 'default',
-  },
+  parameters: { options: { showPanel: false }, a11y: { disable: true }, controls: { disable: true } },
```
You should not disable the controls panel on default story.

```suggestion
  parameters: { a11y: { disable: true } },
```

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.stories.js:
I think I'd be better if you make just one `feed_item_pill_variants.story.vue` template instead of exporting multiple stories depending on default template.

That helps us while visual testing as we just need to take one snapshot and we can see every component variant.

Please refer to: [avatar.stories.js](https://github.com/dialpad/dialtone-vue/blob/staging/components/avatar/avatar.stories.js), [avatar_variants.story.js](https://github.com/dialpad/dialtone-vue/blob/staging/components/avatar/avatar_variants.story.vue) and [avatar.mdx](https://github.com/dialpad/dialtone-vue/blob/staging/components/avatar/avatar.mdx) to get a reference on how should component documentation be written please.

fabiangarga replied:
Got it, definitely this looks better, thanks. Updated the code.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill_constants.js:
```diff
@@ -1,7 +1,12 @@
+  critical: 'dt-recipe-feed-item-pill--border-critical',
+};
+
+export const FEED_ITEM_PILL_DATA_QA = {
+  PILL: 'dt-recipe-feed-item-pill',
+  PILL_ICON: 'dt-recipe-feed-item-pill-icon',
```
Don't export constants that are only used for QA, you can just put the content right in place.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -1,25 +1,25 @@
             :aria-label="ariaLabel"
-            data-qa="dt-feed-item-pill"
-            :class="['d-baw0 d-bgc-moderate d-bar-pill d-w100p d-ta-left d-btn--circle', toggleableClass, buttonClass]"
+            :data-qa="DATA_QA.PILL"
```
We prefer to put the `data-qa` attributes as text don't need to use constants.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -1,25 +1,25 @@
-  <div :class="['dt-feed-item-pill--border', borderClass, wrapperClass]">
-    <div class="d-p8 d-bgc-secondary">
+  <div :class="['dt-recipe-feed-item-pill--border', borderClass, wrapperClass]">
+    <div class="dt-recipe-feed-item-pill--wrapper">
```
```suggestion
    <div class="dt-recipe-feed-item-pill__wrapper">
```

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -1,25 +1,25 @@
-            data-qa="dt-feed-item-pill"
-            :class="['d-baw0 d-bgc-moderate d-bar-pill d-w100p d-ta-left d-btn--circle', toggleableClass, buttonClass]"
+            :data-qa="DATA_QA.PILL"
+            :class="['dt-recipe-feed-item-pill--button', toggleableClass, buttonClass]"
```
```suggestion
            :class="['dt-recipe-feed-item-pill__button', toggleableClass, buttonClass]"
```

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.test.js:
Remove constants please, that's not part of our standards.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -1,25 +1,25 @@
                   size="300"
-                  class="dt-feed-item-pill--icon d-pr8 d-box-content"
-                  data-qa="dt-feed-item-pill-icon"
+                  class="dt-recipe-feed-item-pill--icon"
```
```suggestion
                  class="dt-recipe-feed-item-pill__icon"
```

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -1,25 +1,25 @@
             @click="onClick"
           >
-            <dt-item-layout class="d-w100p d-p8">
+            <dt-item-layout class="dt-recipe-feed-item-pill--layout">
```
```suggestion
            <dt-item-layout class="dt-recipe-feed-item-pill__layout">
```

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -36,7 +36,7 @@
         </template>
         <template #content>
-          <div class="d-jc-center d-d-flex">
+          <div class="dt-recipe-feed-item-pill--content d-jc-center d-d-flex">
```
```suggestion
          <div class="dt-recipe-feed-item-pill__content">
```

juliodialpad reviewed (changes requested):
Gave it another review, please attend comments.

*note*: We're using [Block Element Modifier](https://getbem.com/naming/) naming convention to name our classes.

fabiangarga committed:
feat(DtRecipeFeedItemPill): add variants template

fabiangarga committed:
Merge remote-tracking branch 'origin/staging' into feat_default_state_for_feed_item_pill

fabiangarga commented:
@juliodialpad thanks for the feedback, updated the branch.

fabiangarga committed:
feat(DtRecipeFeedItemPill): remove default from variants

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.stories.js:
```diff
@@ -62,42 +64,28 @@ const DefaultTemplate = (args, { argTypes }) => createTemplateFromVueFile(
-  parameters: Default.parameters,
-  args: {
-    borderColor: 'ai',
+  parameters: {
+    options: { showPanel: false },
+    a11y: { disable: true },
+    percy: {
+      args: { seed: 'seed' },
+    },
```
- The default story should always show the panel, just hide it on the variants story.
- The percy argument `seed` is just for when we're using avatars, so they don't randomly change the color on every visual test, given that, you don't need it on this story.

```suggestion
  parameters: {
    a11y: { disable: true }
  }
```

fabiangarga replied:
Oh makes sense, thought it was added to allow percy to run the visual test for the component, I was not able to access percy so didn't know that. Thanks.

juliodialpad commented on recipes/conversation_view/feed_pill/feed_item_pill.stories.js:
```diff
@@ -62,42 +64,28 @@ const DefaultTemplate = (args, { argTypes }) => createTemplateFromVueFile(
+  render: VariantsTemplate,
+  parameters: {
+    ...Default.parameters,
+    controls: { disable: true },
```
Disable the panel and the controls on variants story.

```suggestion
   parameters: { 
     options: { showPanel: false }, 
     a11y: { disable: true }, 
     controls: { disable: true },
```

juliodialpad reviewed (changes requested):
Previously you were showing `DefaultToggledVariant` and `DefaultToggledVariantNonToggleable` and now I cannot see default toggled examples on variants, I think you missed to add those examples, right?

fabiangarga commented:
> Previously you were showing `DefaultToggledVariant` and `DefaultToggledVariantNonToggleable` and now I cannot see default toggled examples on variants, I think you missed to add those examples, right?

@juliodialpad Correct, due to the conversation here https://dialpad.slack.com/archives/CEDUKPKC1/p1697845368879809
I changed the variants to be more focused in the way we use it in the DP app, by providing some examples of the call/meeting with recaps, which make use of the toggled props

fabiangarga committed:
docs(DtRecipeFeedItemPill): update story params

fabiangarga committed:
Merge remote-tracking branch 'origin/staging' into feat_default_state_for_feed_item_pill

fabiangarga committed:
feat(DtRecipeFeedItemPill): allow title template

fabiangarga commented on recipes/conversation_view/feed_pill/feed_item_pill.vue:
```diff
@@ -1,25 +1,27 @@
-            <dt-item-layout class="d-w100p d-p8">
-              <span class="d-fw-bold">{{ title }}</span>
+            <dt-item-layout class="dt-recipe-feed-item-pill__layout">
+              <slot name="title">
```
@AlexLamDialpad added this so we can use the slot, in case the title needs to be handled differently.

fabiangarga committed:
docs(DtRecipeFeedItemCall): update call/meeting w/recap

Calls/meetings with recaps have the gradient ai as the color of the pill border.

juliodialpad reviewed (approved):
Looking good, thanks!

Please update the branch, add the 'visual-test-ready' tag so visual tests run so I can review and approve them before merging.

juliodialpad committed:
Merge branch 'staging' into feat_default_state_for_feed_item_pill

juliodialpad committed:
add Feed item pill: default story to percy exclusions.

fabiangarga committed:
Merge remote-tracking branch 'origin/staging' into feat_default_state_for_feed_item_pill
